File: editor/canvas.py
import os
from PyQt5.QtCore import Qt, QRectF, pyqtSignal, QTimer
from PyQt5.QtGui import (
    QPainter, QPixmap, QColor, QImage, QBrush,
    QPolygonF, QBitmap, QRegion,
    QPainterPath, QCursor,
)
from PyQt5.QtWidgets import QGraphicsView, QGraphicsScene, QGraphicsPixmapItem

from .layers import LayerStack, Layer
from .history import HistoryManager
from .tools import PencilTool
from .guides import GuideManager
from .snapping import SnappingEngine
from .settings import SettingsManager
from .canvas_drawing import CanvasDrawingMixin
from .canvas_overlay import CanvasOverlayMixin
from .canvas_io import CanvasIOMixin


class CanvasView(CanvasDrawingMixin, CanvasOverlayMixin, CanvasIOMixin, QGraphicsView):
    mouse_moved = pyqtSignal(float, float)
    color_picked = pyqtSignal(QColor)
    bg_color_changed = pyqtSignal(QColor)
    status_changed = pyqtSignal(str)
    zoom_changed = pyqtSignal(float)
    history_changed = pyqtSignal()
    tool_changed = pyqtSignal(str)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.scene = QGraphicsScene(self)
        self.setScene(self.scene)
        self.pixmap_item = QGraphicsPixmapItem()
        self.scene.addItem(self.pixmap_item)

        self.layer_stack = LayerStack(800, 600)
        self.history = HistoryManager()
        self.history.history_changed.connect(self.history_changed.emit)
        self.history.push("New document", self.layer_stack.layers, self.layer_stack.active_index)

        self.zoom_level = 1.0
        self.pan_offset_x = 0
        self.pan_offset_y = 0
        self.tool = PencilTool()
        self.setCursor(getattr(self.tool, 'cursor', QCursor(Qt.ArrowCursor)))
        self.tool_size = 3
        self.tool_color = QColor(0, 0, 0)
        self.tool_opacity = 1.0
        self.tool_flow = 1.0
        self.bg_color = QColor(255, 255, 255)
        self.drawing = False
        self.dragging_layer = False
        self.last_point = None
        self.selection_start = None
        self.selection = None
        self.rubber_band_start = None
        self.rubber_band_end = None
        self.has_rubber_band = False
        self.lasso_points = []
        self.has_lasso = False
        self.gradient_start = None
        self.shape_start = None
        self.clone_source = None

        self.selection_mask = None
        self.selection_path = None
        self.selection_phase = 0
        self.selection_timer = QTimer()
        self.selection_timer.timeout.connect(self._march_selection)
        self.selection_timer.start(120)
        self.guide_mgr = GuideManager()
        self.snapping = SnappingEngine()
        self._settings = SettingsManager()

        self.pen_path = []
        self.pen_handle_offsets = []
        self.vector_paths = []
        self.selected_path_idx = -1
        self.edit_anchor_idx = -1
        self.edit_handle = None
        self.ruler_dragging_guide = False
        self.dragging_guide_index = -1
        self.ruler_drag_orientation = None
        self.ruler_drag_pos = None
        self.crop_active = False
        self.crop_start = None
        self.crop_end = None
        self.crop_drag_handle = None
        self._plugin_filters = {}

        self.show_grid = False
        self.show_rulers = True
        self.snap_indicator = None

        self.setRenderHint(QPainter.Antialiasing)
        self.setRenderHint(QPainter.SmoothPixmapTransform)
        self.setHorizontalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
        self.setVerticalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
        self.setMouseTracking(True)
        self.setDragMode(QGraphicsView.NoDrag)
        self.setViewportUpdateMode(QGraphicsView.FullViewportUpdate)
        self.setAcceptDrops(True)

        self.tool_changed.emit(self.tool.name)
        self._refresh()

    def _refresh(self):
        composite = self.layer_stack.composite()
        self.pixmap_item.setPixmap(QPixmap.fromImage(composite))
        self.scene.setSceneRect(QRectF(composite.rect()))
        self.viewport().update()

    def temp_save_layer(self):
        layer = self.layer_stack.active
        if layer:
            self._temp_layer_image = layer.image.copy()

    def temp_restore_layer(self):
        if hasattr(self, '_temp_layer_image') and self._temp_layer_image:
            layer = self.layer_stack.active
            if layer:
                layer.image = self._temp_layer_image.copy()
                self._refresh()

    def _save_state(self, desc="Edit"):
        self.history.push(desc, self.layer_stack.layers, self.layer_stack.active_index)

    def set_tool(self, tool_name):
        from .tools import TOOLS_BY_NAME
        cls = TOOLS_BY_NAME.get(tool_name.lower())
        if cls:
            self.tool = cls()
            self.setCursor(getattr(self.tool, 'cursor', QCursor(Qt.ArrowCursor)))
            self.tool_changed.emit(self.tool.name)

    def set_tool_size(self, size):
        self.tool_size = max(1, size)

    def set_tool_opacity(self, val):
        self.tool_opacity = val / 100.0

    def set_tool_flow(self, val):
        self.tool_flow = val / 100.0

    def set_foreground_color(self, color):
        self.tool_color = color
        self.color_picked.emit(color)

    def set_background_color(self, color):
        self.bg_color = color
        self.bg_color_changed.emit(color)

    def set_drag_mode(self, mode):
        """0=NoDrag, 1=ScrollHandDrag"""
        self.setDragMode(
            QGraphicsView.ScrollHandDrag if mode == 1 else QGraphicsView.NoDrag
        )
        self.dragging_layer = False

    def zoom_in(self):
        self.zoom_level *= 1.25
        self._apply_zoom()

    def zoom_out(self):
        self.zoom_level /= 1.25
        self._apply_zoom()

    def zoom_fit(self):
        self.fitInView(self.scene.sceneRect(), Qt.KeepAspectRatio)
        self.zoom_level = self.transform().m11()
        self.zoom_changed.emit(self.zoom_level)

    def zoom_100(self):
        self.zoom_level = 1.0
        self._apply_zoom()

    def zoom_to_selection(self):
        if not self.has_selection():
            return
        rect = self.selection_path.boundingRect()
        if rect.isEmpty():
            return
        margin = 20
        rect.adjust(-margin, -margin, margin, margin)
        self.fitInView(rect, Qt.KeepAspectRatio)
        self.zoom_level = self.transform().m11()
        self.zoom_changed.emit(self.zoom_level)

    def _apply_zoom(self):
        self.resetTransform()
        self.scale(self.zoom_level, self.zoom_level)
        self.zoom_changed.emit(self.zoom_level)

    def update_pixmap_position(self):
        pos = self.pixmap_item.pos()
        self.pixmap_item.setPos(pos.x() + self.pan_offset_x, pos.y() + self.pan_offset_y)

    def _march_selection(self):
        if self.selection_mask is not None:
            self.selection_phase = (self.selection_phase + 1) % 6
            self.viewport().update()

    def has_selection(self):
        return self.selection_mask is not None

    def clear_selection(self):
        self.selection_mask = None
        self.selection_path = None
        self.selection_phase = 0
        self.viewport().update()

    def _apply_selection_clip(self, painter):
        if self.selection_mask is None:
            return
        if self.selection_path is not None:
            painter.setClipPath(self.selection_path)
        else:
            mono = self.selection_mask.convertToFormat(QImage.Format_Mono,
                                                       Qt.ThresholdDither)
            bitmap = QBitmap.fromImage(mono)
            region = QRegion(bitmap)
            painter.setClipRegion(region)

    def _selection_mask_from_path(self, path):
        w = int(self.scene.sceneRect().width())
        h = int(self.scene.sceneRect().height())
        mask = QImage(w, h, QImage.Format_ARGB32)
        mask.fill(Qt.transparent)
        p = QPainter(mask)
        p.setRenderHint(QPainter.Antialiasing)
        p.setPen(Qt.NoPen)
        p.setBrush(QBrush(QColor(255, 255, 255)))
        p.drawPath(path)
        p.end()
        return mask

    def set_selection_rect(self, rect):
        from PyQt5.QtCore import QRect
        path = QPainterPath()
        path.addRect(QRectF(rect))
        self.selection_path = path
        self.selection_mask = self._selection_mask_from_path(path)
        self.selection_phase = 0
        self.viewport().update()

    def set_selection_ellipse(self, rect):
        from PyQt5.QtCore import QRect
        path = QPainterPath()
        path.addEllipse(QRectF(rect))
        self.selection_path = path
        self.selection_mask = self._selection_mask_from_path(path)
        self.selection_phase = 0
        self.viewport().update()

    def set_selection_lasso(self, points):
        path = QPainterPath()
        poly = QPolygonF(points)
        path.addPolygon(poly)
        path.closeSubpath()
        self.selection_path = path
        self.selection_mask = self._selection_mask_from_path(path)
        self.selection_phase = 0
        self.viewport().update()

    def set_selection_mask_image(self, mask_qimage):
        self.selection_mask = mask_qimage
        self.selection_path = None
        self.selection_phase = 0
        self.viewport().update()

    def move_layer_content(self, dx, dy):
        layer = self.layer_stack.active
        if not layer or layer.locked:
            return False
        w, h = layer.image.width(), layer.image.height()
        i_dx, i_dy = int(dx), int(dy)
        if i_dx == 0 and i_dy == 0:
            return False
        new_img = QImage(w, h, QImage.Format_ARGB32)
        new_img.fill(Qt.transparent)
        p = QPainter(new_img)
        p.setRenderHint(QPainter.SmoothPixmapTransform)
        p.drawImage(i_dx, i_dy, layer.image)
        p.end()
        layer.image = new_img
        self._refresh()
        return True

    def wheelEvent(self, event):
        if event.modifiers() & Qt.ControlModifier:
            factor = 1.15 if event.angleDelta().y() > 0 else 1 / 1.15
            self.zoom_level *= factor
            self.scale(factor, factor)
            self.zoom_changed.emit(self.zoom_level)
            self.status_changed.emit(f"Zoom: {self.zoom_level * 100:.0f}%")
        else:
            super().wheelEvent(event)

    def dragEnterEvent(self, event):
        if event.mimeData().hasUrls():
            for url in event.mimeData().urls():
                path = url.toLocalFile()
                if path and path.lower().endswith(('.png', '.jpg', '.jpeg', '.bmp', '.gif', '.tiff', '.tif', '.webp')):
                    event.acceptProposedAction()
                    return
        event.ignore()

    def dropEvent(self, event):
        for url in event.mimeData().urls():
            path = url.toLocalFile()
            if not path:
                continue
            ext = os.path.splitext(path)[1].lower()
            if ext not in ('.png', '.jpg', '.jpeg', '.bmp', '.gif', '.tiff', '.tif', '.webp', '.psd'):
                continue
            layer = self.import_image_as_layer(path)
            if layer:
                main = self.window()
                if hasattr(main, 'statusBar'):
                    main.statusBar().showMessage(f"Placed: {os.path.basename(path)}")
                if hasattr(main, 'layer_panel'):
                    main.layer_panel.refresh()
                if hasattr(main, 'nav_panel'):
                    main.nav_panel.refresh()
                event.acceptProposedAction()
                return
        event.ignore()

    def paste_from_clipboard(self):
        from PyQt5.QtWidgets import QApplication
        clipboard = QApplication.clipboard()
        mime = clipboard.mimeData()
        if mime.hasImage():
            img = clipboard.image()
            if img.isNull():
                return False
            if img.format() != QImage.Format_ARGB32:
                img = img.convertToFormat(QImage.Format_ARGB32)
            w, h = self.layer_stack.layers[0].image.width(), self.layer_stack.layers[0].image.height()
            if img.width() != w or img.height() != h:
                img = img.scaled(w, h, Qt.IgnoreAspectRatio, Qt.SmoothTransformation)
            layer = Layer(w, h, "Pasted", Qt.transparent)
            layer.image = img.copy()
            self.layer_stack.layers.append(layer)
            self.layer_stack.active_index = len(self.layer_stack.layers) - 1
            self._save_state("Paste image")
            self._refresh()
            return True
        if mime.hasUrls():
            for url in mime.urls():
                path = url.toLocalFile()
                if path:
                    ext = os.path.splitext(path)[1].lower()
                    if ext in ('.png', '.jpg', '.jpeg', '.bmp', '.gif', '.tiff', '.tif', '.webp', '.psd'):
                        layer = self.import_image_as_layer(path)
                        if layer:
                            self._save_state(f"Paste {os.path.basename(path)}")
                            return True
        return False

    def _ruler_hit_test(self, view_pos):
        ruler_size = 20
        if view_pos.x() < ruler_size:
            return Qt.Vertical
        if view_pos.y() < ruler_size:
            return Qt.Horizontal
        return None

    def mousePressEvent(self, event):
        if event.button() == Qt.RightButton:
            from .tools import PenTool
            if isinstance(self.tool, PenTool):
                self.tool.finalize(self)
            return
        if event.button() == Qt.LeftButton:
            view_pos = event.pos()
            ruler_orientation = self._ruler_hit_test(view_pos)
            if ruler_orientation is not None:
                pos = self.mapToScene(view_pos)
                if not self.guide_mgr.locked:
                    gi = self.guide_mgr.hit_test(pos, 8.0 / self.zoom_level)
                    if gi >= 0:
                        self.dragging_guide_index = gi
                        self.ruler_dragging_guide = True
                        self.ruler_drag_orientation = self.guide_mgr.guides[gi].orientation
                        self.ruler_drag_pos = pos.x() if self.ruler_drag_orientation == Qt.Vertical else pos.y()
                    else:
                        self.ruler_dragging_guide = True
                        self.ruler_drag_orientation = ruler_orientation
                        self.ruler_drag_pos = pos.x() if ruler_orientation == Qt.Vertical else pos.y()
                return
            self.drawing = True
            pos = self.mapToScene(view_pos)
            sn_pos, snap_info = self._snap_point(pos)
            self.last_point = sn_pos
            mods = int(event.modifiers())
            self._save_state(f"{self.tool.name if hasattr(self.tool, 'name') else 'Tool'}")
            self.tool.press(self, sn_pos, mods)
            if snap_info:
                self._show_snap_indicator(snap_info[0], snap_info[1])
                self.status_changed.emit(f"Snap to {snap_info[1]}")
        elif event.button() == Qt.MiddleButton:
            self.setDragMode(QGraphicsView.ScrollHandDrag)
            super().mousePressEvent(event)

    def mouseMoveEvent(self, event):
        pos = self.mapToScene(event.pos())
        self.mouse_moved.emit(pos.x(), pos.y())
        if self.ruler_dragging_guide:
            self.ruler_drag_pos = pos.x() if self.ruler_drag_orientation == Qt.Vertical else pos.y()
            self.viewport().update()
            return
        if self.drawing and self.last_point:
            mods = int(event.modifiers())
            sn_pos, snap_info = self._snap_point(pos)
            self.tool.move(self, self.last_point, sn_pos, mods)
            self.last_point = sn_pos
            if snap_info:
                self._show_snap_indicator(snap_info[0], snap_info[1])
                self.status_changed.emit(f"Snap to {snap_info[1]}")
        super().mouseMoveEvent(event)

    def mouseReleaseEvent(self, event):
        if event.button() == Qt.LeftButton:
            if self.ruler_dragging_guide:
                pos = self.mapToScene(event.pos())
                if self.dragging_guide_index >= 0:
                    if not self.guide_mgr.locked:
                        view_pos = event.pos()
                        ruler_orientation = self._ruler_hit_test(view_pos)
                        if ruler_orientation is not None:
                            self.guide_mgr.remove_guide(self.dragging_guide_index)
                        else:
                            g = self.guide_mgr.guides[self.dragging_guide_index]
                            g.position = pos.x() if g.orientation == Qt.Vertical else pos.y()
                    self.dragging_guide_index = -1
                else:
                    pv = pos.x() if self.ruler_drag_orientation == Qt.Vertical else pos.y()
                    if pv >= 0:
                        self.guide_mgr.add_guide(self.ruler_drag_orientation, pv)
                self.ruler_dragging_guide = False
                self.ruler_drag_orientation = None
                self.ruler_drag_pos = None
                self.viewport().update()
                return
            if self.drawing:
                self.drawing = False
                mods = int(event.modifiers())
                pos = self.mapToScene(event.pos())
                sn_pos, snap_info = self._snap_point(pos)
                self.tool.release(self, sn_pos, mods)
                if snap_info:
                    self._show_snap_indicator(snap_info[0], snap_info[1])
                    self.status_changed.emit(f"Snap to {snap_info[1]}")
        elif event.button() == Qt.MiddleButton:
            self.setDragMode(QGraphicsView.NoDrag)
            event.ignore()

    def keyPressEvent(self, event):
        key = event.key()
        mods = event.modifiers()

        # Tool shortcuts (Photoshop-style)
        if key == Qt.Key_V and not mods: self.set_tool("Move Tool")
        elif key == Qt.Key_M and not mods: self.set_tool("Rectangular Marquee Tool")
        elif key == Qt.Key_L and not mods: self.set_tool("Lasso Tool")
        elif key == Qt.Key_W and not mods: self.set_tool("Magic Wand Tool")
        elif key == Qt.Key_B and not mods: self.set_tool("Brush Tool")
        elif key == Qt.Key_P and not mods: self.set_tool("Pen Tool")
        elif key == Qt.Key_N and not mods: self.set_tool("Pencil Tool")
        elif key == Qt.Key_E and not mods: self.set_tool("Eraser Tool")
        elif key == Qt.Key_G and not mods: self.set_tool("Gradient Tool")
        elif key == Qt.Key_U and not mods: self.set_tool("Rectangle Tool")
        elif key == Qt.Key_S and not mods: self.set_tool("Clone Stamp Tool")
        elif key == Qt.Key_J and not mods: self.set_tool("Spot Healing Brush Tool")
        elif key == Qt.Key_C and not mods: self.set_tool("Crop Tool")
        elif key == Qt.Key_T and not mods: self.set_tool("Horizontal Type Tool")
        elif key == Qt.Key_H and not mods: self.set_tool("Hand Tool")
        elif key == Qt.Key_Z and not mods: self.set_tool("Zoom Tool")
        elif key == Qt.Key_I and not mods: self.set_tool("Eyedropper Tool")
        elif key == Qt.Key_K and not mods: self.set_tool("Paint Bucket Tool")

        elif key == Qt.Key_BracketRight and not mods: self.set_tool_size(self.tool_size + 1)
        elif key == Qt.Key_BracketLeft and not mods: self.set_tool_size(max(1, self.tool_size - 1))

        # Enter / Return to finalize pen or apply crop
        elif key in (Qt.Key_Return, Qt.Key_Enter) and not mods:
            from .tools import PenTool, CropTool
            if isinstance(self.tool, PenTool):
                self.tool.finalize(self)
            elif isinstance(self.tool, CropTool):
                self.tool.apply(self)

        # Backspace to delete selected anchor in pen tool
        elif key == Qt.Key_Backspace and not mods:
            from .tools import PenTool
            if isinstance(self.tool, PenTool):
                self.tool.key_press(self, key)

        # Escape to cancel pen or crop
        elif key == Qt.Key_Escape and not mods:
            from .tools import PenTool, CropTool
            if isinstance(self.tool, PenTool):
                self.tool.cancel(self)
            elif isinstance(self.tool, CropTool):
                self.tool.cancel(self)

        # Ctrl+
        elif mods & Qt.ControlModifier:
            if key == Qt.Key_Z:
                if mods & Qt.ShiftModifier:
                    if self.history.can_redo():
                        self.history.redo(self.layer_stack)
                        self._refresh()
                else:
                    if self.history.can_undo():
                        self.history.undo(self.layer_stack)
                        self._refresh()
            elif key == Qt.Key_V:
                if self.paste_from_clipboard():
                    self.status_changed.emit("Pasted image from clipboard")
                    main = self.window()
                    if hasattr(main, 'layer_panel'):
                        main.layer_panel.refresh()
                    if hasattr(main, 'nav_panel'):
                        main.nav_panel.refresh()
            elif key == Qt.Key_Plus: self.zoom_in()
            elif key == Qt.Key_Minus: self.zoom_out()
            elif key == Qt.Key_0: self.zoom_fit()
            elif key == Qt.Key_1: self.zoom_100()
            elif key == Qt.Key_2: self.zoom_to_selection()

        super().keyPressEvent(event)

File: cpp_editor/src/canvas.cpp
#include "canvas.h"
#include "tools.h"
#include <QWheelEvent>
#include <QMouseEvent>
#include <QKeyEvent>
#include <QPainter>
#include <QFileInfo>
#include <queue>
#include <set>
#include <tuple>

CanvasView::CanvasView(QWidget *parent)
    : QGraphicsView(parent)
{
    scene_ = new QGraphicsScene(this);
    setScene(scene_);
    pixmapItem_ = new QGraphicsPixmapItem();
    scene_->addItem(pixmapItem_);

    layerStack = LayerStack(800, 600);

    auto shots = snapshotFromLayers(layerStack.layers);
    history.push("New document", shots, layerStack.activeIndex_);

    setRenderHint(QPainter::Antialiasing);
    setRenderHint(QPainter::SmoothPixmapTransform);
    setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    setMouseTracking(true);
    setDragMode(QGraphicsView::ScrollHandDrag);
    tool = new PencilTool();
    refresh();
}

void CanvasView::refresh()
{
    QImage composite = layerStack.composite();
    pixmapItem_->setPixmap(QPixmap::fromImage(composite));
    scene_->setSceneRect(composite.rect());
}

void CanvasView::saveState(const QString &desc)
{
    auto shots = snapshotFromLayers(layerStack.layers);
    history.push(desc, shots, layerStack.activeIndex_);
}

void CanvasView::newImage(int width, int height, const QColor &bg)
{
    layerStack = LayerStack(width, height);
    if (bg != Qt::white)
        layerStack.layers[0].image.fill(bg);
    history.clear();
    auto shots = snapshotFromLayers(layerStack.layers);
    history.push("New document", shots, layerStack.activeIndex_);
    refresh();
}

bool CanvasView::openImage(const QString &path)
{
    QImage img(path);
    if (img.isNull()) return false;
    if (img.format() != QImage::Format_ARGB32)
        img = img.convertToFormat(QImage::Format_ARGB32);

    layerStack = LayerStack(img.width(), img.height());
    layerStack.layers[0].image = img.copy();
    history.clear();
    auto shots = snapshotFromLayers(layerStack.layers);
    history.push("Open " + QFileInfo(path).fileName(), shots, layerStack.activeIndex_);
    refresh();
    return true;
}

bool CanvasView::saveImage(const QString &path)
{
    QImage composite = layerStack.composite();
    return composite.save(path);
}

bool CanvasView::exportPng(const QString &path)
{
    QImage composite = layerStack.composite();
    return composite.save(path, "PNG");
}

void CanvasView::setTool(const QString &toolName)
{
    delete tool;
    if (toolName == "pencil") tool = new PencilTool();
    else if (toolName == "brush") tool = new BrushTool();
    else if (toolName == "eraser") tool = new EraserTool();
    else if (toolName == "color_picker") tool = new ColorPickerTool();
    else if (toolName == "flood_fill") tool = new FloodFillTool();
    else tool = new PencilTool();
}

void CanvasView::drawPoint(const QPointF &pos)
{
    auto layer = layerStack.active();
    if (!layer || layer->locked) return;

    QPainter painter(&layer->image);
    painter.setRenderHint(QPainter::Antialiasing);
    QPen pen(toolColor_, toolSize_, Qt::SolidLine, Qt::RoundCap, Qt::RoundJoin);
    painter.setPen(pen);
    painter.drawPoint(pos);
    painter.end();
    refresh();
}

void CanvasView::drawLine(const QPointF &p1, const QPointF &p2)
{
    auto layer = layerStack.active();
    if (!layer || layer->locked) return;

    QPainter painter(&layer->image);
    painter.setRenderHint(QPainter::Antialiasing);
    QPen pen(toolColor_, toolSize_, Qt::SolidLine, Qt::RoundCap, Qt::RoundJoin);
    painter.setPen(pen);
    painter.drawLine(p1, p2);
    painter.end();
    refresh();
}

void CanvasView::erasePoint(const QPointF &pos)
{
    auto layer = layerStack.active();
    if (!layer || layer->locked) return;

    QPainter painter(&layer->image);
    painter.setRenderHint(QPainter::Antialiasing);
    painter.setCompositionMode(QPainter::CompositionMode_Clear);
    QPen pen(QColor(0, 0, 0, 0), toolSize_, Qt::SolidLine, Qt::RoundCap, Qt::RoundJoin);
    painter.setPen(pen);
    painter.drawPoint(pos);
    painter.end();
    refresh();
}

void CanvasView::eraseLine(const QPointF &p1, const QPointF &p2)
{
    auto layer = layerStack.active();
    if (!layer || layer->locked) return;

    QPainter painter(&layer->image);
    painter.setRenderHint(QPainter::Antialiasing);
    painter.setCompositionMode(QPainter::CompositionMode_Clear);
    QPen pen(QColor(0, 0, 0, 0), toolSize_, Qt::SolidLine, Qt::RoundCap, Qt::RoundJoin);
    painter.setPen(pen);
    painter.drawLine(p1, p2);
    painter.end();
    refresh();
}

QColor CanvasView::getPixelColor(const QPointF &pos) const
{
    auto layer = layerStack.active();
    if (!layer) return QColor();
    int x = qBound(0, (int)pos.x(), layer->image.width() - 1);
    int y = qBound(0, (int)pos.y(), layer->image.height() - 1);
    return layer->image.pixelColor(x, y);
}

void CanvasView::floodFill(const QPointF &pos)
{
    auto layer = layerStack.active();
    if (!layer || layer->locked) return;

    int x = (int)pos.x(), y = (int)pos.y();
    int w = layer->image.width(), h = layer->image.height();
    if (x < 0 || x >= w || y < 0 || y >= h) return;

    QColor target = layer->image.pixelColor(x, y);
    if (target == toolColor_) return;

    std::queue<std::pair<int,int>> queue;
    std::set<std::tuple<int,int>> visited;
    queue.push({x, y});
    visited.insert({x, y});

    QPainter painter(&layer->image);
    painter.setPen(QPen(toolColor_, 1));

    while (!queue.empty()) {
        auto [cx, cy] = queue.front(); queue.pop();
        if (layer->image.pixelColor(cx, cy) == target) {
            painter.drawPoint(cx, cy);
            for (auto [nx, ny] : {std::pair{cx+1,cy}, {cx-1,cy}, {cx,cy+1}, {cx,cy-1}}) {
                if (nx >= 0 && nx < w && ny >= 0 && ny < h && !visited.count({nx, ny})) {
                    visited.insert({nx, ny});
                    queue.push({nx, ny});
                }
            }
        }
    }
    painter.end();
    refresh();
}

void CanvasView::zoomIn()
{
    zoomLevel_ *= 1.25;
    resetTransform();
    scale(zoomLevel_, zoomLevel_);
}

void CanvasView::zoomOut()
{
    zoomLevel_ /= 1.25;
    resetTransform();
    scale(zoomLevel_, zoomLevel_);
}

void CanvasView::zoomFit()
{
    fitInView(scene_->sceneRect(), Qt::KeepAspectRatio);
    zoomLevel_ = transform().m11();
}

void CanvasView::zoom100()
{
    zoomLevel_ = 1.0;
    resetTransform();
}

void CanvasView::wheelEvent(QWheelEvent *event)
{
    if (event->modifiers() & Qt::ControlModifier) {
        double factor = event->angleDelta().y() > 0 ? 1.15 : 1.0 / 1.15;
        zoomLevel_ *= factor;
        scale(factor, factor);
        emit statusChanged(QString("Zoom: %1%").arg((int)(zoomLevel_ * 100)));
    } else {
        QGraphicsView::wheelEvent(event);
    }
}

void CanvasView::mousePressEvent(QMouseEvent *event)
{
    if (event->button() == Qt::LeftButton) {
        drawing_ = true;
        QPointF pos = mapToScene(event->pos());
        lastPoint_ = pos;
        saveState(tool ? tool->name() : "Edit");
        if (tool) tool->press(this, pos);
    }
}

void CanvasView::mouseMoveEvent(QMouseEvent *event)
{
    QPointF pos = mapToScene(event->pos());
    emit mouseMoved(pos.x(), pos.y());
    if (drawing_ && tool) {
        tool->move(this, lastPoint_, pos);
        lastPoint_ = pos;
    }
}

void CanvasView::mouseReleaseEvent(QMouseEvent *event)
{
    if (event->button() == Qt::LeftButton && drawing_) {
        drawing_ = false;
        if (tool) tool->release(this, mapToScene(event->pos()));
    }
}

void CanvasView::keyPressEvent(QKeyEvent *event)
{
    if (event->modifiers() & Qt::ControlModifier) {
        if (event->key() == Qt::Key_Z) {
            if (event->modifiers() & Qt::ShiftModifier) {
                if (history.canRedo()) {
                    auto shots = snapshotFromLayers(layerStack.layers);
                    history.redo(shots, layerStack.activeIndex_);
                    restoreLayers(layerStack.layers, shots);
                    refresh();
                }
            } else {
                if (history.canUndo()) {
                    auto shots = snapshotFromLayers(layerStack.layers);
                    history.undo(shots, layerStack.activeIndex_);
                    restoreLayers(layerStack.layers, shots);
                    refresh();
                }
            }
            return;
        }
        if (event->key() == Qt::Key_Plus) { zoomIn(); return; }
        if (event->key() == Qt::Key_Minus) { zoomOut(); return; }
        if (event->key() == Qt::Key_0) { zoomFit(); return; }
    }
    QGraphicsView::keyPressEvent(event);
}
